alacritty/src/display/damage.rs

```diff
@@ -0,0 +1,345 @@
+use alacritty_terminal::vi_mode::ViModeCursor;
+
+/// Minimum part of damaged square when intersecting damaged rectangles.
+pub const MIN_DAMAGED_SQUARE_PART: f32 = 0.4;
+
+/// An error when we for look ahead in iterator, used to try to collect rectangles which pretty
+/// close to each other, but don't fit into `MIN_DAMAGED_SQUARE_PART` just by looking on the first
+/// next damaged rectangle.
+pub const MIN_DAMAGED_SQUARE_PART_ERR: f32 = MIN_DAMAGED_SQUARE_PART / 2.;
+
+pub struct RendererDamage<'a> {
+    index: usize,
+    rect: Option<Rect>,
+    size_info: SizeInfo<u32>,
+    terminal_damage: &'a mut Vec<DamagedLine>,
+}
+
+impl<'a> RendererDamage<'a> {
+    pub fn new(
+        terminal_damage: &'a mut TermDamage,
+        size_info: SizeInfo<u32>,
+        cursor: RenderableCursor,
+        vi_cursor: Option<ViModeCursor>,
+        selection: Option<SelectionRange>,
+        display_offset: usize,
+    ) -> Self {
+        let fully_damaged = terminal_damage.take_fully_damaged();
+        let terminal_damage =
+            terminal_damage.damage_mut(cursor, vi_cursor, selection, display_offset);
+
+        let (index, rect) = if fully_damaged {
+            let rect = Rect { x: 0, y: 0, width: size_info.width(), height: size_info.height() };
+            (terminal_damage.len(), rect)
+        } else {
+            // Terminal cursor is always damaged there's at least one damaged line.
+            let index = terminal_damage.iter().position(|&line| line.is_damaged()).unwrap();
+            let rect = Self::rect_for_line(&size_info, index, terminal_damage[index]);
+
+            terminal_damage[index].reset();
+
+            // Advance index into the next unprocessed element.
+            (index + 1, rect)
+        };
+
+        Self { index, rect: Some(rect), terminal_damage, size_info }
+    }
+
+    /// Creates a new rectangle in a way that 'lhs' and 'rhs' are inside of it.
+    #[inline]
+    fn intersect(lhs: Rect, rhs: Rect) -> Rect {
+        let left_x = cmp::min(lhs.x, rhs.x);
+        let right_x = cmp::max(lhs.x + lhs.width, rhs.x + rhs.width);
+        let y_top = cmp::max(lhs.y + lhs.height, rhs.y + rhs.height);
+        let y_bottom = cmp::min(lhs.y, rhs.y);
+        Rect { x: left_x, y: y_bottom, width: right_x - left_x, height: y_top - y_bottom }
+    }
+
+    #[inline]
+    fn rect_for_line(size_info: &SizeInfo<u32>, line: usize, line_damage: DamagedLine) -> Rect {
+        let x = size_info.padding_x() + line_damage.left as u32 * size_info.cell_width();
+        let y = size_info.height()
+            - size_info.padding_y()
+            - (line + 1) as u32 * size_info.cell_height();
+        let width = (line_damage.right - line_damage.left + 1) as u32 * size_info.cell_width();
+        Rect { x, y, height: size_info.cell_height(), width }
+    }
+
+    /// Adds extra damage area around the `rect` to account for glyphs that are larger than a glyph
+    /// cell or weird wide characters.
```

Comment:
Should use imperative.

```suggestion
    /// Add extra damage area around the rect to account for glyphs that are larger than their cell.
```

I'd assume that by `or weird wide characters` you mean wide characters extending beyond their cell, not that wide characters require over-damage to be rendered correctly? I'd assume the terminal always damages wide characters like two halfwidth.

Reply:
It actually doesn't damage full width chars properly. It was doing that initially, but then I've removed that part to speed up things. So right now we do that to include wide chars.